lib/mutgen.c

```diff
@@ -25,66 +25,350 @@
+    // additions to tsk_mutation_t:
+    double time;
+    struct mutation_t *prev;
+    struct mutation_t *next;
+    bool new;
+} mutation_t;
+
 typedef struct {
-    const char *ancestral_state;
-    const char *derived_state;
-} mutation_type_t;
+    double position;
+    char *ancestral_state;
+    tsk_size_t ancestral_state_length;
+    char *metadata;
+    tsk_size_t metadata_length;
+    // modifications to tsk_site_t:
+    mutation_t *mutations;
+    size_t mutations_length;
+    bool new;
+} site_t;
+
+static int
+cmp_site(const void *a, const void *b)
+{
+    const site_t *ia = (const site_t *) a;
+    const site_t *ib = (const site_t *) b;
+    return (ia->position > ib->position) - (ia->position < ib->position);
+}
+
+static int
+cmp_mutationp(const void *a, const void *b)
+{
+    int out;
+    /* the extra cast is to avoid a gcc bug in -Werror=cast-qual:
+     * https://gcc.gnu.org/bugzilla/show_bug.cgi?id=81631 */
+    const mutation_t *ia = *(const mutation_t **) (uintptr_t) a;
+    const mutation_t *ib = *(const mutation_t **) (uintptr_t) b;
+    out = ia->id < 0 || ib->id < 0 ? 0 : (ia->id - ib->id) - (ib->id - ia->id);
+    if (out == 0)
+        out = (ib->time > ia->time) - (ib->time < ia->time);
+    if (out == 0)
+        out = (ia->parent > ib->parent) - (ia->parent < ib->parent);
+    return out;
+}
 
-static const mutation_type_t binary_mutation_types[] = {
-    {"0", "1"}
+static void 
+insert_mutation(site_t *site, mutation_t *new)
+{
+    mutation_t *u;
+    u = site->mutations;
+    new->prev = NULL;
+    new->next = u;
+    if (u != NULL) {
+        u->prev = new;
+    }
+    site->mutations = new;
+    site->mutations_length++;
+    assert(site->mutations_length > 0);
+}
+
+static void 
+delete_mutation(site_t *site, mutation_t *mut)
+{
+    mutation_t *u;
+    // TODO: need to dealloc mut?
+    u = mut->prev;
+    if (u == NULL) {
+        site->mutations = mut->next;
+    } else {
+        u->next = mut->next;
+    }
+    site->mutations_length--;
+}
+
+static void
+sort_mutations(site_t *site)
+{
+    size_t k;
+    mutation_t *m;
+    mutation_t *p[site->mutations_length];
+    size_t num_mutations = site->mutations_length;
+    k = 0;
+    for (m = site->mutations; m != NULL; m = m->next) {
+        p[k] = m;
+        k++;
+    }
+    assert(k == num_mutations);
+    qsort(p, (size_t) num_mutations, sizeof(mutation_t *), &cmp_mutationp);
+    site->mutations = p[0];
+    for (k = 0; k < num_mutations; k++) {
+        m = p[k];
+        if (k == 0) {
+            m->prev = NULL;
+        } else {
+            m->prev = p[k-1];
+        }
+        if (k == num_mutations - 1) {
+            m->next = NULL;
+        } else {
+            m->next = p[k+1];
+        }
+    }
+}
+
+/*******************
+ * Mutation model */
+
+/* We will use these as dummy values, that should all be replaced. */
+static const char dummy_alphabet[2] = "XY";
+
+// TODO: why do alphabets have const but other things not?
```

Comment:
Any time you use a string constant it has to be assigned to a const char * variable. This can get *really* tedious.